File: Python/algospot/JLIS.py
import sys
input = lambda: sys.stdin.readline().rstrip() # 속도 최적화 (효과있나?)

C = int(input())

memo = dict()

results = []

for _ in range(C):
    memo.clear()

    a, b = tuple(map(int, input().split()))

    numA = [0] + list(map(int, input().split()))
    numB = [0] + list(map(int, input().split()))

    def solve(indexA, indexB): # int, int -> int
        if indexA > a or indexB > b:
            return 0

        if (indexA, indexB) not in memo:
            valA = numA[indexA]
            valB = numB[indexB]
    
            result = 0
    
            for newIndexA in range(indexA + 1, a + 1):
                newVal = numA[newIndexA]
                if newVal > valA and newVal > valB:
                    result = max(result, solve(newIndexA, indexB))
    
            for newIndexB in range(indexB + 1, b + 1):
                newVal = numB[newIndexB]
                if newVal > valA and newVal > valB:
                    result = max(result, solve(indexA, newIndexB))

            memo[(indexA, indexB)] = result

        return memo[(indexA, indexB)] + 1

    result = 0
    for i in range(-1, a + 1):
        for j in range(-1, b + 1):
            result = max(result, solve(i, j) - 1)

    results.append(str(result))

print('\n'.join(results))

'''
print('--- result ---')

i = 1
for result in results:
    print('result #' + str(i) + ': ' + str(result))
    i += 1
'''

File: CPP/JLIS.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cstring>

std::vector<int> numA, numB;
int lenA, lenB, num;

int memo[100+1][100+1];

int solve(int indexA, int indexB) {
    if ((indexA > lenA) || (indexB > lenB))
        return 0;

    int valA = numA.at(indexA);
    int valB = numA.at(indexB);

    if (memo[indexA][indexB] == -1) {
        int result = 0;

        for (int newIndexA = indexA + 1; newIndexA <= lenA; newIndexA++) {
            int newVal = numA.at(newIndexA);
            if ((newVal > valA) && (newVal > valB))
                result = std::max(result, solve(newIndexA, indexB));
        }

        for (int newIndexB = indexB + 1; newIndexB <= lenB; newIndexB++) {
            int newVal = numB.at(newIndexB);
            if ((newVal > valA) && (newVal > valB))
                result = std::max(result, solve(indexA, newIndexB));
        }

        memo[indexA][indexB] = result;
    }

    return memo[indexA][indexB] + 1;
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int N;
    std::cin >> N;
    for (int i = 0; i < N; i++) {
        numA.clear();
        numB.clear();
        std::cin >> lenA >> lenB;

        for (int i = 0; i < lenA; i++) {
            std::cin >> num;
            numA.push_back(num);
        }
        
        for (int i = 0; i < lenB; i++) {
            std::cin >> num;
            numB.push_back(num);
        }

        memset(memo, -1, sizeof(memo));

        int result = 0;

        for (int i = 0; i <= lenA; i++)
            for (int j = 0; j <= lenB; j++)
                result = std::max(result, solve(i, j) - 1);

        std::cout << result << std::endl; // 랙?
    }

    return 0;
}